lib/pure/asyncdispatch.nim

```diff
@@ -355,27 +355,50 @@ proc `or`*[T, Y](fut1: Future[T], fut2: Future[Y]): Future[void] =
   fut2.callback = cb
   return retFuture
 
-proc all*[A](futs: seq[Future[A]]): Future[seq[A]] =
+proc all*[T](futs: varargs[Future[T]]): auto =
   ## Returns a future which will complete once all futures in ``futs``
   ## complete.
   ##
   ## The resulting future will hold the values of all awaited futures,
   ## in the order they are passed.
 
-  var
-    retFuture = newFuture[seq[A]]("asyncdispatch.all")
-    retValues = newSeq[A](len(futs))
-    completedFutures = 0
+  when T is void:
+    var
+      retFuture = newFuture[void]("asyncdispatch.all")
+      completedFutures = 0
 
-  for i, fut in futs:
-    fut.callback = proc(f: Future[A]) =
-      retValues[i] = f.read()
-      inc(completedFutures)
+    let totalFutures = len(futs)
 
-      if completedFutures == len(futs):
-        retFuture.complete(retValues)
+    for i, fut in futs:
+      proc setCallback(i: int) =
```

Comment:
The body of this proc doesn't seem to use `i` so there should be no need for this `setCallback`.